api/agent/agent.go

```diff
@@ -334,52 +341,52 @@ func (a *agent) hotLauncher(ctx context.Context, callObj *call) {
 
 		curStats := callObj.slots.getStats()
 		isNeeded := isNewContainerNeeded(&curStats)
-		logger.WithFields(logrus.Fields{
-			"currentStats": curStats,
-			"isNeeded":     isNeeded,
-		}).Debug("Hot function launcher stats")
+		logger.WithFields(logrus.Fields{"currentStats": curStats, "isNeeded": isNeeded}).Debug("Hot function launcher stats")
 		if !isNeeded {
 			continue
 		}
+		logger.WithFields(logrus.Fields{"currentStats": curStats, "isNeeded": isNeeded}).Info("Hot function launcher starting hot container")
 
-		ctxResource, cancelResource := context.WithCancel(context.Background())
-		logger.WithFields(logrus.Fields{
-			"currentStats": curStats,
-			"isNeeded":     isNeeded,
-		}).Info("Hot function launcher starting hot container")
+		// if we get a timeout, we must call off get resource token
+		ctx, cancel := context.WithCancel(ctx)
 
 		select {
-		case tok, isOpen := <-a.resources.GetResourceToken(ctxResource, callObj.Memory, uint64(callObj.CPUs), isAsync):
-			cancelResource()
+		case tok, isOpen := <-a.resources.GetResourceToken(ctx, callObj.Memory, uint64(callObj.CPUs), isAsync):
+			cancel()
 			if isOpen {
 				a.wg.Add(1)
 				go func(ctx context.Context, call *call, tok ResourceToken) {
 					a.runHot(ctx, call, tok)
 					a.wg.Done()
 				}(ctx, callObj, tok)
```

Comment:
I think ctx here should be different, we are passing a cancelled ctx to runHot, right?

Reply:
we probably should to reduce confusion, but we're really just extracting out the span in `runHot`

Reply:
gonna fiddle with this block